lib/models/room.js

```diff
@@ -284,54 +546,12 @@ Room.prototype.addEventsToTimeline = function(events, toStartOfTimeline) {
             }
         }
 
-        setEventMetadata(events[i], stateContext, toStartOfTimeline);
-        // modify state
-        if (events[i].isState()) {
-            stateContext.setStateEvents([events[i]]);
-            // it is possible that the act of setting the state event means we
-            // can set more metadata (specifically sender/target props), so try
-            // it again if the prop wasn't previously set. It may also mean that
-            // the sender/target is updated (if the event set was a room member event)
-            // so we want to use the *updated* member (new avatar/name) instead.
-            if (!events[i].sender || events[i].getType() === "m.room.member") {
-                setEventMetadata(events[i], stateContext, toStartOfTimeline);
-            }
-        }
-        if (events[i].getType() === "m.room.redaction") {
-            // try to remove the element
-            var removed = utils.removeElement(
-                this.timeline, checkForRedaction(events[i])
-            );
-            if (!removed && toStartOfTimeline) {
-                // redactions will trickle in BEFORE the event redacted so make
-                // a note of the redacted event; we'll check it later.
-                this._redactions.push(events[i].event.redacts);
-            }
-            // NB: We continue to add the redaction event to the timeline so clients
-            // can say "so and so redacted an event" if they wish to.
-        }
+        var spliceBeforeLocalEcho = !isLocalEcho && addLocalEchoToEnd;
 
-        // TODO: pass through filter to see if this should be added to the timeline.
-        if (toStartOfTimeline) {
-            this.timeline.unshift(events[i]);
-        }
-        else {
-            // everything gets added to the end by default. What we actually want to
-            // do in this scenario is *NOT* add REAL events to the end if there are
-            // existing local echo events at the end.
-            if (addLocalEchoToEnd && !isLocalEcho) {
-                var insertIndex = this.timeline.length;
-                for (var j = this.timeline.length - 1; j >= 0; j--) {
-                    if (!this.timeline[j].status) { // real events don't have a status
-                        insertIndex = j + 1;
-                        break;
-                    }
-                }
-                this.timeline.splice(insertIndex, 0, events[i]); // insert element
-            }
-            else {
-                this.timeline.push(events[i]);
-            }
+        if (!this._eventIdToTimeline[events[i].getId()]) {
```

Comment:
Presumably this dictionary is going to grow to be an unbounded map of all the event IDs we've ever seen. How would we go about purging this?